source/common/network/io_socket_handle_impl.cc

```diff
@@ -280,6 +280,37 @@ absl::optional<uint32_t> maybeGetPacketsDroppedFromHeader([[maybe_unused]] const
   return absl::nullopt;
 }
 
+template <typename T> absl::optional<T> maybeGetUnsignedIntFromHeader(const cmsghdr& cmsg) {
+  static_assert(std::is_unsigned_v<T>, "return type must be unsigned integral");
+  switch (cmsg.cmsg_len) {
+  case CMSG_LEN(sizeof(uint8_t)):
+    return static_cast<T>(*reinterpret_cast<const uint8_t*>(CMSG_DATA(&cmsg)));
+  case CMSG_LEN(sizeof(uint16_t)):
+    return static_cast<T>(*reinterpret_cast<const uint16_t*>(CMSG_DATA(&cmsg)));
+  case CMSG_LEN(sizeof(uint32_t)):
+    return static_cast<T>(*reinterpret_cast<const uint32_t*>(CMSG_DATA(&cmsg)));
+  case CMSG_LEN(sizeof(uint64_t)):
+    return static_cast<T>(*reinterpret_cast<const uint64_t*>(CMSG_DATA(&cmsg)));
```

Comment:
I think that unless the CMSG data happens to be correctly aligned, this may cause an unaligned access error, so we should use memcpy instead. Could we do:
```
  T value;
  if (cmsg.cmsg_len != CMSG_LEN(sizeof(value)) {
    IS_ENVOY_BUG(
      fmt::format("unexpected cmsg_len value for unsigned integer payload: {}", cmsg.cmsg_len));
  return absl::nullopt;
  }
  memcpy(&value, CMSG_DATA(&cmsg), sizeof(value));
  return value;
```

Reply:
The conditional wouldn't work because `(cmsg.cmsg_len != CMSG_LEN(sizeof(value))` is usually true because `cmsg.cmsg_len` is usually greater than `CMSG_LEN(sizeof(value))`. The `memcpy` would work on little endian like it does now but it would still have the same problems on big endian. See https://github.com/envoyproxy/envoy/pull/37217#issuecomment-2491427032 for details. 

I don't know enough about data alignment to know if its a problem here but I couldn't find any other places in envoy that used `memcpy` on `CMSG_DATA(&cmsg)` so maybe that should be addressed separately?

Reply:
Oh, actually a colleague of mine found a reference to the alignment issue:
https://man7.org/linux/man-pages/man3/cmsg.3.html

```
CMSG_DATA()
              returns a pointer to the data portion of a cmsghdr.  The
              pointer returned cannot be assumed to be suitably aligned
              for accessing arbitrary payload data types.  Applications
              should not cast it to a pointer type matching the payload,
              but should instead use memcpy(3) to copy data to or from a
              suitably declared object.
```

So let's go ahead and memcpy() into a temporary object of the right type based on the length and then